server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ResizeAllocationDecider.java

```diff
@@ -31,16 +30,15 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocat
 
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
-        final UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
-        if (unassignedInfo != null && shardRouting.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
+        if (shardRouting.unassigned() && shardRouting.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
```

Comment:
Just looking into whether this is equivalent, I'm unsure whether it's possible to have a `ShardRouting` which is assigned but which has `unassignedInfo != null`? We check `assert (state == ShardRoutingState.UNASSIGNED && unassignedInfo == null) == false` in its constructor but maybe we should `assert (state == ShardRoutingState.UNASSIGNED) == (unassignedInfo != null)` instead?